bpf: Migrate `cilium-migrate-map` from C to Go

nathanjsweet opened the pull request:
Currently, initializing bpf programs and migrating maps is done by a C program implementing an ELF parser. This patch introduces a Go version of the program where `cilium/ebpf` does the heavy lifting instead. The feature is now also exposed as a Go API, making it callable from other parts of the agent. This version also parses and verifies the ELF's BTF section.

A CLI wrapper is temporarily included to make it callable from init.sh while it is translated to Go.

Best reviewed per commit.

```release-note
bpf: Migrate map migration logic from C to Go
```

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,306 @@
+// 	MapSecIdx elf.SectionIndex
+// }
+
+// type bpfELfMap struct {
```
Did you modify the library to read `bpf_elf_map` definitions correctly? Out of the box, only `bpf_map_def` is supported, which has the fields up to `flags`. Then the iproute2 `bpf_elf_map` adds the `id` and `pinning` fields, and the CIlium fork adds the `innerFD` and `innerIdx` fields.

If we want `cilium/ebpf` to read these map definitions correctly, we'd have to add some API to parse trailing bytes out-of-band, or add the capability to supply a custom unmarshaler for ELF map definitions. Alternatively, we could try and switch to BTF map definitions first (#15149), but not sure where we are wrt. iproute2 fork compatibility. @brb @borkmann

nathanjsweet replied:
Hmmm. My impression was that for the purposes of migration those fields are actually not important. They get equalized when the comparison is made. I don't think this program actually needs those fields.

brb replied:
> but not sure where we are wrt. iproute2 fork compatibility

I'm still working on upstreaming patches to the kernel (all have been submitted, just needs to be accepted:). Anyway, if it takes longer than anticipated, I will just backport the patches to our iproute2 and libbpf forks and will make the PR ready for the review. Another anyway, is that the PR (with backported patches) has been passing the CI: https://github.com/cilium/cilium/pull/16727.

I think the BTF map definition change is blocked (commented in the issue).

ti-mo commented on cilium/cmd/bpf_map_migrate.go:
```diff
@@ -0,0 +1,73 @@
+			}
+		},
+	}
+	s string
```
These are package-level vars, so should probably be named a bit more descriptively.

ti-mo commented on cilium/cmd/bpf_map_migrate.go:
```diff
@@ -0,0 +1,73 @@
+				defer sysLogger.Close()
+			}
+			if !finalize {
+				_, err = migrate.Start(pathName, sysLogger)
```
Should we do something with the return value? It's always discarded.

nathanjsweet replied:
So the next step is to get rid of the shell script altogether using Louis' new manager code:
https://github.com/cilium/cilium/compare/cilium:fc6ef4d...ldelossa:louis/golang-init
For that manager code the boolean value will be helpful to skipping the `Finish` function.

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+)
+
+func init() {
+	fsBase = filepath.Join(defaults.DefaultMapRoot, "/tc/globals")
```
Could this be moved to the var declaration, or is DefaultMapRoot not populated yet at that point?

nathanjsweet replied:
Good call, this is vestigial from when there was more logic, it can be moved.

ti-mo commented on bpf/init.sh:
```diff
@@ -260,12 +260,12 @@ function bpf_load()
 	tc qdisc replace dev $DEV clsact || true
 	[ -z "$(tc filter show dev $DEV $WHERE | grep -v 'pref 1 bpf chain 0 $\|pref 1 bpf chain 0 handle 0x1')" ] || tc filter del dev $DEV $WHERE
-	cilium-map-migrate -s $OUT
+        cilium map migrate -s $OUT
```
A wild indentation appeared~

ti-mo commented on cilium/cmd/bpf_map_migrate.go:
```diff
@@ -0,0 +1,73 @@
+			sysLogger, err := syslog.New(syslog.LOG_WARNING, "cilium-map-migrate")
+			if err != nil {
+				log.Printf("Could not open syslog: %v", err)
+				sysLogger = nil
```
Should already be nil if err != nil?

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+	statePending = "pending"
+	// cf https://elixir.bootlin.com/linux/latest/source/samples/bpf/tc_l2_redirect_kern.c#L22
+	pinGlobalNS = 2
+	utimeNow    = (1 << 30) - 1
```
Where does this value come from?

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+	if syslogger != nil {
+		syslogger.Warning(fmt.Sprintf("Property mismatch in %s, migrating node to %s!\n", file, dest))
+	}
+	err = syscall.UtimesNano(file, nilTimeSpecArg)
```
Why does this need to be done? We're bumping the timestamp, but not really checking it anywhere. Can we get away without it?

cc @borkmann

nathanjsweet replied:
I'm fairly certain we want this for kernel debugging purposes.

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+	if err != nil {
+		return false, fmt.Errorf("could not update timestamp of %q", file)
+	}
+	err = os.Rename(file, dest)
```
Why isn't `pinned.Pin()` used here? It was specifically implemented for this migration use case. :slightly_smiling_face: It safely re-pins existing pinned maps, too.

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+	if syslogger != nil {
+		syslogger.Warning(fmt.Sprintf("Unlinking migrated node %s due to good exit.\n", file))
+	}
+	return syscall.Unlink(file)
```
What's the difference with `os.Remove()`?

nathanjsweet replied:
In case there is another link to the inode. `os.Delete` deletes the inode out-right.

ti-mo replied:
Sorry, meant to say `os.Remove()`, that one seems to call `syscall.Unlink()` under the covers, I think that's preferred to using `syscall` here.

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+		}
+		return fmt.Errorf("cannot stat node %q", file)
+	}
+	if exit != 0 {
```
I'd invert the control flow here with an early return so most of the function body doesn't need to live indented unnecessarily.

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+		if err != nil {
+			return fmt.Errorf("could not update timestamp of %q", file)
+		}
+		err := unix.Renameat2(unix.AT_FDCWD, file, unix.AT_FDCWD, dest, 1)
```
I'd use `LoadPinnedMap()` followed by a `m.Pin()` here as well like I suggested above.

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+		if syslogger != nil {
+			syslogger.Warning(fmt.Sprintf("Restoring migrated node %s into %s due to bad exit.\n", file, dest))
+		}
+		err = syscall.UtimesNano(file, nilTimeSpecArg)
```
Same as my earlier comment, why do acc and mod timestamps need to be bumped? Is this something we should incorporate into `Map.Pin()`?

nathanjsweet replied:
Potentially. I'd be interested in a more comprehensive from @borkmann .

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+	}
+	var moved bool
+	for n, m := range coll.Maps {
+		var pinning uint32
```
Parsing the extra bytes could go into a separate function to keep things clear.

ti-mo commented on pkg/bpf/migrate/migrate.go:
```diff
@@ -0,0 +1,376 @@
+	}
+	return moved, nil
+}
+
```
```suggestion
// bpfPending attempts to load a pinned map with the given name from `.../tc/globals`
// and compares its type, key size, value size, flags and max entries to the given MapSpec.
// If any of the attributes differ, re-pin the Map to its current name prefixed by `:pending`.
```

ti-mo commented on bpf/init.sh:
```diff
@@ -260,12 +260,12 @@ function bpf_load()
 	RETCODE=$?
 	set -e
-	cilium-map-migrate -e $OUT -r $RETCODE
+	cilium map migrate -e $OUT -r $RETCODE
```
Could we call `cilium map migrate` with the correct parameters instead of simply passing `$?` into a Go program? :/ This makes for spaghetti downstream.

nathanjsweet replied:
My take is that we will be getting rid of this anyways. ASAP. This PR is only to ensure that the core migrate logic works as intended.

nathanjsweet replied:
My take is that we will be getting rid of this anyways. ASAP. This PR is only to ensure that the core migrate logic works as intended.

ti-mo reviewed (commented):
Hey, thanks for taking this on. I think most of the code around renaming map pins can be removed, as the lib has `Map.Pin()` that was introduced specifically for this renaming use case.

I don't think we should pass in `$RETCODE`, let's handle any conditionals in the script itself.

Also, we should take #16561 into account, but can be done in a follow-up. Will further the discussion on the issue itself.

ti-mo committed:
vendor: upgrade cilium/ebpf to v0.7.0

Upgrade the lib to make use of new pinning features.

Signed-off-by: Timo Beckers <timo@isovalent.com>

nathanjsweet committed:
cilium-migrate-map: port from from C to Go

Currently initializing bpf programs and migrating maps is done by a C program
implementing an ELF parser. This patch introduces a Go version of the program
where cilium/ebpf does the heavy lifting instead. The feature is now also
exposed as a Go API, making it callable from other parts of the agent.
This version also parses and verifies the ELF's BTF section.

A CLI wrapper is temporarily included to make it callable from init.sh while
it is translated to Go.

Signed-off-by: Nate Sweet <nathanjsweet@pm.me>
Co-authored-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
cilium-map-migrate: remove program and all of its references

This has been replaced with a Go-based implementation, so no longer needs to
be part of the build/packaging process.

The CILIUM_BPF_MNT env has been removed from shell invocations pertaining to
bpffs since it was only referenced in cilium-map-migrate.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
loader: replaceQdisc/Datapath - perform bpffs map migrations using Go

This uses the newly-ported Go logic to execute bpffs map migrations in
replaceQdisc and replaceDatapath.

Also simplified the control flow a bit to avoid accidentally shadowing err.
The revert behaviour has been made more explicit.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
elf: use errors.Is() in symbol extraction, simplify logic

Document the ELF parsing logic to make it easier to grok.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
elf: patch ELF chunks using Seek() instead of tracking global offset

This patch uses io.WriteSeeker.Seek() to navigate to each symbol's offset
instead of alternating between streaming input from the original ELF and
inscribing replacement values.

The new approach is slightly more straightforward as no global offset tracking
needs to take place.

Signed-off-by: Timo Beckers <timo@isovalent.com>

ti-mo committed:
elf: perform string substitutions in .BTF section

The way we currently perform string substitutions in ELFs no longer holds up
with the presence of a BTF blob. The .BTF section has its own string table
for representing type metadata, and cilium/ebpf uses this table to perform
lookups for building its internal representations of the BTF type graph.
This behaviour should not be altered, as it makes sure all ELF symbols are
correctly represented in the BTF blob.

This patch extends our current approach to also rewrite pieces of the .BTF
section, which will be required for as long as we keep using a mix of iproute2
and cilium/ebpf for reading ELFs.

Signed-off-by: Timo Beckers <timo@isovalent.com>

tklauser commented on pkg/elf/symbols.go:
```diff
@@ -189,13 +210,13 @@ func (s *symbols) extractFrom(e *elf.File) error {
 		// off is the offset within the string table.
 		elfString, err = stringReader.ReadString(nullTerminator)
-		if err != nil && err != io.EOF {
+		if errors.Is(err, io.EOF) {
```
Looks like this accidentially got inverted:

```suggestion
		if !errors.Is(err, io.EOF) {
```

tklauser commented on pkg/elf/symbols.go:
```diff
@@ -214,11 +235,27 @@ func (s *symbols) extractFrom(e *elf.File) error {
+		// Read BTF header from the .BTF section.
+		h, err := readBTFHeader(btfSec, e.ByteOrder)
+		if err != nil {
+			return fmt.Errorf("reading BTF section header: %s", err)
```
```suggestion
			return fmt.Errorf("reading BTF section header: %w", err)
```

tklauser commented on pkg/elf/symbols.go:
```diff
@@ -214,11 +235,27 @@ func (s *symbols) extractFrom(e *elf.File) error {
+
+		// Populate offsetBTF fields of all symbols.
+		if err := findBTFSymbols(stringOffsets, btfSec, h); err != nil {
+			return fmt.Errorf("finding BTF string offsets: %s", err)
```
```suggestion
			return fmt.Errorf("finding BTF string offsets: %w", err)
```

tklauser commented on pkg/elf/btf.go:
```diff
@@ -0,0 +1,102 @@
+
+	// Seek to the location of the BTF string table.
+	if _, err := rs.Seek(int64(btfStrTabOff), io.SeekStart); err != nil {
+		return fmt.Errorf("seeking to BTF string table: %s", err)
```
```suggestion
		return fmt.Errorf("seeking to BTF string table: %w", err)
```

tklauser commented on pkg/elf/btf.go:
```diff
@@ -0,0 +1,102 @@
+			break
+		}
+		if err != nil {
+			return fmt.Errorf("reading BTF string table: %s", err)
```
```suggestion
			return fmt.Errorf("reading BTF string table: %w", err)
```

tklauser reviewed (commented):
Some drive-by nits, haven't looked at it in detail yet.

tklauser commented on bpf/cilium-map-migrate.c:
```diff
@@ -1,576 +0,0 @@
-
-#include <linux/bpf.h>
-
-#include "elf/libelf.h"
-#include "elf/gelf.h"
-
-#include "bpf/ctx/nobpf.h"
```
It looks like `cilium-map-migrate.c` is the only user of the headers in `bpf/include/elf` and the `bpf/ctx/nobpf.h`. So I think we can safely remove these as part of this PR as well.

Also, we no longer need to link binaries against `libelf` so we can remove `-lelf` from `bpf/Makefile` here:

https://github.com/cilium/cilium/blob/6a722d3234b6f31237f2def6739092b19a7ccc24/bpf/Makefile#L306-L307

ti-mo replied:
Thanks, that's great news! :grin: Will tackle in a follow-up. :+1:

tklauser reviewed (approved):
Thanks @ti-mo for walking me through the change. LGTM with one minor comment which can be done in a follow-up PR if needed.

pchaigno commented on pkg/bpf/bpffs_migrate.go:
```diff
@@ -0,0 +1,188 @@
+		return nil
+	}
+
+	log.Infof("Unpinning map '%s' after successful recreation", file)
```
```suggestion
	log.WithField(logfields.BPFMapPath, file).Info("Unpinning map after successful recreation")
```

pchaigno commented on pkg/bpf/bpffs_migrate.go:
```diff
@@ -0,0 +1,188 @@
+	// Pending Map was found on bpffs and needs to be reverted.
+	if revert {
+		dest := filepath.Join(bpffsPath, name)
+		log.Infof("Reverting map pin from '%s' to '%s' after failed migration", file, dest)
```
```suggestion
		log.WithField(logfields.BPFMapPath, dest).Infof("Repinning without '%s' suffix after failed migration", bpffsPending)
```

pchaigno commented on pkg/bpf/bpffs_migrate.go:
```diff
@@ -0,0 +1,188 @@
+
+	dest := file + bpffsPending
+
+	log.Infof("New version of map '%s' has different properties, re-pinning from '%s' to '%s'", name, file, dest)
```
```suggestion
	log.WithFields({logfields.BPFMapName: name, logfields.BPFMapPath: file}).Infof("New version of map '%s' has different properties, re-pinning with '%s' suffix", bpffsPending)
```

pchaigno commented on pkg/datapath/loader/netlink.go:
```diff
@@ -109,35 +108,26 @@ func replaceDatapath(ctx context.Context, ifName, objPath, progSec, progDirectio
-		} else {
-			retCode = "1"
+		if err := bpf.FinalizeBPFFSMigration(bpf.GetMapRoot(), objPath, revert); err != nil {
+			log.WithError(err).WithField("mapPath", mapPath).WithField("objPath", objPath).
```
```suggestion
			log.WithError(err).WithFields({logfields.BPFMapPath: mapPath, "objPath": objPath}).
```

pchaigno reviewed (commented):
Nice stuff! The pull request is really easy to review thanks to all the comments :-)

I have a couple nits below, which I'll send as a separate PR given I'm a bit late to the merge party.
